code/modules/vehicles/hardpoints/primary/arc_sentry.dm

```diff
@@ -0,0 +1,231 @@
+		if(living_mob.stat == DEAD)
+			purge_target(living_mob)
+			continue
+
+		if(living_mob.get_target_lock(faction_group) || living_mob.invisibility || HAS_TRAIT(living_mob, TRAIT_ABILITY_BURROWED))
+			purge_target(living_mob)
+			continue
+
+		var/list/turf/path = get_line(get_turf(src), living_mob)
+		if(!length(path) || get_dist(get_turf(src), living_mob) > turret_range)
+			purge_target(living_mob)
+			continue
+
+		var/blocked = FALSE
+		for(var/turf/tile as anything in path)
+			if(tile.density || tile.opacity)
+				blocked = TRUE
+				break
+
+			for(var/obj/structure/struct in tile)
+				if(struct.opacity)
+					blocked = TRUE
+					break
+
+			for(var/obj/vehicle/multitile/vehicle in tile)
+				if(vehicle == owner) // Some of the tiles will inevitably be the ARC itself
+					continue
+				blocked = TRUE
+				break
+
+			if(locate(/obj/effect/particle_effect/smoke) in tile)
+				blocked = TRUE
+				break
+
+		if(blocked)
+			purge_target(living_mob)
+			continue
+
+		if(living_mob.stat & UNCONSCIOUS)
+			unconscious_targets += living_mob
+		else
+			conscious_targets += living_mob
+
+	if((sentry_target in conscious_targets) || (sentry_target in unconscious_targets))
+		sentry_target = sentry_target
+
+	else if(length(conscious_targets))
+		sentry_target = pick(conscious_targets)
+
+	else if(length(unconscious_targets))
+		sentry_target = pick(unconscious_targets)
+
+	if(!sentry_target) //No targets, don't bother firing
+		return
+
+	start_fire(object = sentry_target)
+
+/obj/item/hardpoint/primary/arc_sentry/can_be_removed(mob/remover)
+	var/obj/vehicle/multitile/arc/arc_owner = owner
+	if(!istype(arc_owner))
+		return TRUE
+
+	if(arc_owner.antenna_deployed)
+		to_chat(remover, SPAN_WARNING("[src] cannot be removed from [owner] while its antenna is deployed."))
+		return FALSE
+
+	return ..()
+
+/obj/projectile/arc_sentry/Initialize(mapload, datum/cause_data/cause_data)
+	. = ..()
+	RegisterSignal(src, COMSIG_BULLET_POST_HANDLE_OBJ, PROC_REF(check_passthrough))
+
+/obj/projectile/arc_sentry/check_canhit(turf/current_turf, turf/next_turf)
+	var/proj_dir = get_dir(current_turf, next_turf)
+	var/obj/item/hardpoint/arc_sentry = shot_from
+	if(!(arc_sentry.owner in current_turf) && !(arc_sentry.owner in next_turf) && (proj_dir & (proj_dir - 1)) && !current_turf.Adjacent(next_turf))
+		ammo.on_hit_turf(current_turf, src)
+		current_turf.bullet_act(src)
+		return TRUE
+
+	// Check for hits that would occur when moving to turf, such as a blocking cade
+	if(scan_a_turf(next_turf, proj_dir))
+		return TRUE
+
+	return FALSE
+
+/obj/projectile/arc_sentry/proc/check_passthrough(datum/source, obj/hit_obj, bool)
+	SIGNAL_HANDLER
+
+	if(!istype(shot_from, /obj/item/hardpoint))
+		return
+
+	var/obj/item/hardpoint/sentry = shot_from
+	if(sentry.owner == hit_obj)
+		return COMPONENT_BULLET_PASS_THROUGH
+
+/obj/item/hardpoint/primary/arc_sentry/proc/purge_target(mob/target)
+	if(target == sentry_target)
+		sentry_target = null
+	targets.Remove(target)
```

Comment:
This would make more sense up with the other `/obj/item/hardpoint/primary/arc_sentry` procs (maybe just below `get_target()`). Currently it's split off from them by `/obj/projectile/arc_sentry`.

Reply:
done